SceneView : Add FPS counter

danieldresser-ie opened the pull request:
Fairly basic version, but it seems to do the trick for checking the performance of scene manipulators.

FPSGadget may also be added to the GraphGadget for monitoring that, though I haven't added any UI for doing this.

johnhaddon commented on include/GafferSceneUI/SceneView.h:
```diff
@@ -139,6 +140,8 @@ class GAFFERSCENEUI_API SceneView : public GafferUI::View
 		class Gnomon;
 		std::unique_ptr<Gnomon> m_gnomon;
+		class FPS;
+		std::unique_ptr<FPS> m_fps;
```
This is an ABI break, so we should target `main` rather than `0.61_maintenance`.

johnhaddon commented on src/GafferUI/FPSGadget.cpp:
```diff
@@ -0,0 +1,115 @@
+		return;
+	}
+
+	static std::chrono::time_point<std::chrono::high_resolution_clock> lastTime;
```
What happens when we make two different FPSGadgets? Doesn't this being static cause problems there?

danieldresser-ie replied:
Oh, yep, I should fix that.

johnhaddon commented on src/GafferUI/FPSGadget.cpp:
```diff
@@ -0,0 +1,115 @@
+	}
+
+	static std::chrono::time_point<std::chrono::high_resolution_clock> lastTime;
+	static bool initialized = false;
```
Perhaps `std::optional<time_point>`? That makes it impossible to use an uninitialised value, and is more explicit than a separate `initialized` variable (which could refer to anything).

johnhaddon commented on src/GafferUI/FPSGadget.cpp:
```diff
@@ -0,0 +1,115 @@
+	const ViewportGadget *viewportGadget = ancestor<ViewportGadget>();
+	ViewportGadget::RasterScope rasterScope( viewportGadget );
+
+	glTranslatef( 5, 50, 0.0f );
```
This seems like a reasonable position for the Viewer, but less so for the GraphEditor (or other potential usage points). Should we accessors to allow it to be specified? Or maybe we should just use `getTransform()` so that `setTransform()` can be used to provide an arbitrary raster-space transform?

danieldresser-ie replied:
Lets see, I don't see any need to change it dynamically, could probably just be a constructor argument, defaulting to this position?

danieldresser-ie replied:
Hmm, that's fair.  My thought was just that I would prefer to have it show up somewhere reasonable if you don't override the transform.  Is it reasonable to have a constructor argument with a default value that calls setTransform?

johnhaddon replied:
> My thought was just that I would prefer to have it show up somewhere reasonable if you don't override the transform.
 
 Fair point - I guess the identity transform would put it in the top left, but rendering just outside of frame?
 
 > Is it reasonable to have a constructor argument with a default value that calls setTransform?

The whole subject of raster space gadgets is a bit of an odd one really. I think maybe a more principled approach would be to move the RasterScope out of the Gadgets themselves, and have the ViewportGadget manage separate containers of raster and non-raster gadgets, organising the appropriate transform scope for each before drawing. Then FPSGadget could report a proper bound, and we'd just put it in a standard container gadget that would be responsible for performing the layout. That's a whole other topic though...

Passing a transform to the constructor would definitely be an oddity as far as other gadgets go, but it seems like a reasonable compromise here...

johnhaddon commented on src/GafferUI/FPSGadget.cpp:
```diff
@@ -0,0 +1,115 @@
+	glTranslatef( 5, 50, 0.0f );
+	glScalef( 10.0f, -10.0f, 10.0f );
+
+	style->renderText( Style::LabelText, boost::str( boost::format( "%.1f FPS" ) % ( 1.0f / averageFrameTime ) ) );
```
This font/size seems more prominent than other more important aspects of the UI. Could we try BodyText and/or a slightly smaller font?

danieldresser-ie replied:
I didn't find BodyText very readable, dropping from a scale of 10 to 8 looks pretty decent though.

johnhaddon commented on src/GafferUI/FPSGadget.cpp:
```diff
@@ -0,0 +1,115 @@
+	}
+
+	// Simple running average
+	m_totalTime += frameTime;
+	m_timeBuffer.push_back( frameTime );
+
+	// 20 frames is enough to smooth out bumps, while not lagging behind too much
+	while( m_timeBuffer.size() > 20 )
+	{
+		m_totalTime -= m_timeBuffer.front();
+		m_timeBuffer.pop_front();
+	}
+
+	float averageFrameTime = m_totalTime / ( m_timeBuffer.size() * 1000000.0f );
```
I've tended to use an exponential moving average for this sort of thing in the past, because it doesn't need a buffer : https://stackoverflow.com/a/4687507. Is there any downside to that?

danieldresser-ie replied:
My thought was that it's important to have a definite period of time over which the old values are definitely completely gone, particularly because unlike in a lot of things with FPS counters, it's possible to have extremely slow updates in Gaffer ( because nothing may trigger an update for an extended period of time ).

The buffer size of 20 is actually a pretty awkward compromise between stability and updating quickly enough when the FPS is low.

I realized last night that I could do something much more reliable: average over the past 1 second.  If the update has taken longer than a second, then the FPS only considers the last frame.  I think my idea for implementing that may look a little cleaner anyway.

johnhaddon replied:
> it's possible to have extremely slow updates in Gaffer ( because nothing may trigger an update for an extended period of time )

I did have one thought on that but forgot to mention it on the PR. Perhaps we could/should uninitialise `lastTime` on `visibilityChangedSignal()` so that an extended period of not being drawn due to invisibility doesn't throw off the initial readouts? Your point would still stand about periods where the gadget is visible but not needing a redraw for other reasons though.

> I realized last night that I could do something much more reliable: average over the past 1 second. If the update has taken longer than a second, then the FPS only considers the last frame. I think my idea for implementing that may look a little cleaner anyway.

Cool, sounds worth a try. Let's try to get this wrapped up quickly though - I think any of the approaches is probably sufficient for our purposes, and there's plenty else we need to move on to.

johnhaddon reviewed (changes requested):
Thanks Daniel, this is going to be handy. Comments inline as usual...

danieldresser-ie commented:
I've addressed all feedback.  I've gone with the completely different way of computing FPS, which seems to work quite nicely - it seems quite helpful when waiting for 2 seconds to get an update produces an immediate drop to 0.5 FPS, rather than having to wait for several slow updates before it converges.  That invalidates some of the specific comments, about static and `initialized`, but I think it addresses the overall concerns.
